### PR TITLE
Libusb fixups

### PR DESCRIPTION
**Pull request by shuffle2**
fixes some regressions from updating libusb on macos

### CODE REVIEW

**Comment by pizuz** [comment]
Again, no luck. Dolphin can't access the adapter. Master does.

Screenshot 2022-04-22 at 17 31 43

**Comment by pizuz** [comment]
Scratch that. I think, something's wrong on my end. Master should recognize the adapter, but doesn't ATM. I'm investigating.

**Comment by pizuz** [comment]
For some reason, the buildbot build of this PR doesn't log anything. The newest log I found in ~/Library/Application Support/Dolphin/Logs was from 2014. Do I have to build it myself, then?

**Comment by shuffle2** [comment]
nah that's ok, i've just disabled the detach call on macos now...so it should work. LMK if you still see the crash/hang tho

**Comment by pizuz** [comment]
Alright. Sorry, but I have close to no software development experience (aside from some C64 BASIC and Assembler and some Pascal from high school).

Anyway, this build works as intended. The Adapter is recognized and the freeze on quit has also been fixed. Thank you for looking into it. Maybe someone else will be able to find a definitive solution for this. Connecting a GCAdapter without all those hoops you have to jump through on macOS seems quite desirable.

File: Source/Core/InputCommon/GCAdapter.cpp [hunk 1]
```diff
@@ -71,7 +80,7 @@ static bool s_libusb_hotplug_enabled = true;
 #else
 static bool s_libusb_hotplug_enabled = false;
 #endif
-#if defined(LIBUSB_API_VERSION) && LIBUSB_API_VERSION >= 0x01000102
+#if LIBUSB_API_VERSION_ATLEAST(0x01000102)
```
Comment (AdmiralCurtiss) [inline Source/Core/InputCommon/GCAdapter.cpp:83]: I assume you wanted `#if LIBUSB_API_HAS_HOTPLUG` here?

**Comment by pizuz** [comment]
Still working as intended :)

**Comment by ryanmcgrath** [comment]
I'll add one more time that the bug you were experiencing occurs even with older versions of libusb, like what's used over in Slippi Dolphin.

I'm not entirely sure this is related to anything in newer libusb, since those changes in libusb/`detach_kernel_driver` shouldn't be doing anything anyway without being run as root and/or having the right entitlement. Am I missing something...?

**Review by AdmiralCurtiss — APPROVED** [review]
Untested, and can't say much about the macOS thing, but looks good to me.

**Comment by pizuz** [comment]
> I'll add one more time that the bug you were experiencing occurs even with older versions of libusb, like what's used over in Slippi Dolphin.
> 
> I'm not entirely sure this is related to anything in newer libusb, since those changes in libusb/`detach_kernel_driver` shouldn't be doing anything anyway without being run as root and/or having the right entitlement. Am I missing something...?

There were two bugs that have been fixed here. One freeze on exit introduced in #10502 and the GCAdapter not being recognized introduced in #10577.

**Comment by ryanmcgrath** [comment]
>the GCAdapter not being recognized introduced in https://github.com/dolphin-emu/dolphin/pull/10577.

I know. I am telling you that this bug exhibits in older forks of Dolphin and I don't believe it was _introduced_ in a recent build; I regularly help users debug this issue.

>The detach feature should work the other way round. The GCAdapter is always captured by a system driver and libusb cannot access it, unless you install a [codeless KEXT/DEXT which prevents such capture.](https://secretkeys.io/gcadapterdriver/)

I know. I'm the author of that driver.

>That new libusb feature would free the device for complete access, thus making those codeless KEXTs and DEXTs obsolete. However, it seems like there are additional steps necessary in order for this to work in Dolphin.

This requires either an entitlement that Apple doesn't seem to hand out, or running under `sudo` - the libusb feature is somewhat useless without it, so I don't quite see how it'd work out in the end here.

**Comment by pizuz** [comment]
Sorry for not recognizing… Thanks for that driver, by the way. It‘s been working very reliably, at least on my Catalina box.

Anyway, at least this particular instance could be tracked down to exactly that feature introduced in libusb 1.0.25. My bisect was [here](https://github.com/dolphin-emu/dolphin/pull/10577#issuecomment-1106346879). What‘s causing the issues in the other forks, I don‘t know because I use none of them.

**Comment by ryanmcgrath** [comment]
>Thanks for that driver, by the way. It‘s been working very reliably, at least on my Catalina box.

Glad to hear it!

>Anyway, at least this particular instance could be tracked down to exactly that feature introduced in libusb 1.0.25. My bisect was https://github.com/dolphin-emu/dolphin/pull/10577#issuecomment-1106346879. What‘s causing the issues in the other forks, I don‘t know because I use none of them.

Well, I shall eat my words.

I see now that [libusb on macOS moved to returning](https://github.com/libusb/libusb/blob/ba698478afc3d3a72644eef9fc4cd24ce8383a4c/libusb/os/darwin_usb.c#L2535-L2579) `LIBUSB_ERROR_ACCESS`, whereas before it just [always returned](https://github.com/dolphin-emu/dolphin/blob/8466d438d3675657fa997983d27d06d804d500b1/Externals/libusb/libusb/os/darwin_usb.c#L1683-L1687) `LIBUSB_ERROR_NOT_SUPPORTED`... which was special-cased here. 

This was helpful to see, since now it makes me realize that `LIBUSB_ERROR_ACCESS` in the older forks seems to come up moreso due to some issue with `kIOReturnExclusiveAccess`. I can debug that there with this info at least.

**Comment by pizuz** [comment]
> > Thanks for that driver, by the way. It‘s been working very reliably, at least on my Catalina box.
> 
> Glad to hear it!
> 
> > Anyway, at least this particular instance could be tracked down to exactly that feature introduced in libusb 1.0.25. My bisect was [#10577 (comment)](https://github.com/dolphin-emu/dolphin/pull/10577#issuecomment-1106346879). What‘s causing the issues in the other forks, I don‘t know because I use none of them.
> 
> Well, I shall eat my words.
> 
> I see now that [libusb on macOS moved to returning](https://github.com/libusb/libusb/blob/ba698478afc3d3a72644eef9fc4cd24ce8383a4c/libusb/os/darwin_usb.c#L2535-L2579) `LIBUSB_ERROR_ACCESS`, whereas before it just [always returned](https://github.com/dolphin-emu/dolphin/blob/8466d438d3675657fa997983d27d06d804d500b1/Externals/libusb/libusb/os/darwin_usb.c#L1683-L1687) `LIBUSB_ERROR_NOT_SUPPORTED`... which was special-cased here.
> 
> This was helpful to see, since now it makes me realize that `LIBUSB_ERROR_ACCESS` in the older forks seems to come up moreso due to some issue with `kIOReturnExclusiveAccess`. I can debug that there with this info at least.

I still have that previous build which [hadn't patched out libusb_detach_kernel_driver](https://github.com/dolphin-emu/dolphin/pull/10599/commits/12cd81bdb3e9d72f8a9abb430342ab9532e7dab1), in case you want it to investigate this further.

**Comment by AdmiralCurtiss** [comment]
This is pretty simple and straightfoward, so I'll just merge this so USB stuff works on macOS again.

**Pull request merged**